src/macports1.0/migrate.tcl

```diff
@@ -37,16 +37,57 @@ package require restore 1.0
 package require selfupdate 1.0
 
 namespace eval migrate {
+    ##
+    # The main function. Calls each individual step in order.
+    #
+    # @returns 0 on success, -999 when MacPorts base has been upgraded and the
+    #          caller should re-run itself and invoke migration with the --continue
+    #          flag set.
     proc main {opts} {
-        # The main function. Calls each individual step in order.
-        #
-        # Args:
-        #           opts - options array.
-        # Returns:
-        #           0 if success
-
         array set options $opts
 
+        if {[needs_migration]} {
+            if {[info exists macports::ui_options(questions_yesno)]} {
+                set msg "Migration will first upgrade MacPorts and then reinstall all installed ports."
+                set retvalue [$macports::ui_options(questions_yesno) $msg "MigrationPrompt" "" {y} 0 "Would you like to continue?"]
+                if {$retvalue == 1} {
+                    # quit as user answered 'no'
+                    ui_msg "Aborting migration. You can re-run 'sudo port migrate' later or follow the migration instructions: https://trac.macports.org/wiki/Migration"
+                    return 0
+                }
+            }
+
+            ui_msg "Upgrading MacPorts..."
+            set success no
+            if {[catch {set success [upgrade_port_command]} result]} {
+                ui_debug $::errorInfo
+                ui_error "Upgrading port command failed. Try running 'sudo port -v selfupdate' and then 'sudo port migrate'."
+                return 1
+            }
+            if {!$success} {
+                ui_error "Upgrading port command failed or was not attempted. Please re-install MacPorts manually and then run 'sudo port migrate' again."
+                return 1
+            }
+
+            # MacPorts successfully upgraded, automatically re-run migration
```

Comment:
Okay, this looks something which I wasn't able to figure out at that time. I had this problem that how do I run the new port command just installed, so I broke it into two parts. First one creates a snapshot and upgrades, and the second one restores.

I think you are trying to do the same here but without breaking the procedure and instead re-running the migration and using two different return flags. Am I right?

This will change the design of restore a bit though.

Reply:
Yes, I've made it part of the API of `migrate::main` that callers must re-execute themselves and run `migrate::main` with `options(ports_migrate_continue)` set to do this. `port(1)` (implemented in `src/port/port.tcl`) does this by using the Tcl extension `execl` from TclX (documented in `TclX(n)`).

We use a similar trick for selfupdate already; if selfupdate did update MacPorts, it returns `-999` and the interactive port interpreter (that you'd get by running `port(1)` without any arguments) will quit. We could actually apply the same `execl` method there – at the time when this code was written, we did not have access to the Tcl `execl` extension, though.